python/pyiceberg/io/pyarrow.py

```diff
@@ -532,131 +535,116 @@ def project_table(
+        return struct_result
 
-    def _get_field_by_id(self, field_id: int) -> Optional[NestedField]:
-        try:
-            return self.file_schema.find_field(field_id)
-        except ValueError:
-            # Field is not in the file
+    def struct(
+        self, struct: StructType, struct_array: Optional[pa.Array], field_results: List[Optional[pa.Array]]
+    ) -> Optional[pa.Array]:
+        if struct_array is None:
             return None
 
-    @visit_with_arrow.register(StructType)
-    def _(self, requested_struct: StructType, file_struct: Optional[IcebergType]) -> pa.Array:  # pylint: disable=unused-argument
-        """Visit a StructType with a concrete SchemaVisitorWithPartner"""
-        results = []
-
-        for requested_field in requested_struct.fields:
-            file_field = self._get_field_by_id(requested_field.field_id)
-
-            if file_field is None and requested_field.required:
-                raise ResolveException(f"Field is required, and could not be found in the file: {requested_field}")
-
-            results.append(self.visit_with_arrow(requested_field.field_type, file_field))
-
-        pyarrow_schema = schema_to_pyarrow(requested_struct)
-        return pa.StructArray.from_arrays(arrays=results, fields=pyarrow_schema)
-
-    @visit_with_arrow.register(ListType)
-    def _(self, requested_list: ListType, file_field: Optional[NestedField]) -> pa.Array:
-        """Visit a ListType with a concrete SchemaVisitorWithPartner"""
+        field_arrays: List[pa.Array] = []
+        fields: List[pa.Field] = []
+        for pos, field_array in enumerate(field_results):
+            field = struct.fields[pos]
+            if field_array is not None:
+                array = self.cast_if_needed(field, field_array)
+                field_arrays.append(array)
+                fields.append(pa.field(field.name, array.type, field.optional))
+            elif field.optional:
+                arrow_type = schema_to_pyarrow(field.field_type)
+                field_arrays.append(pa.nulls(len(struct_array)).cast(arrow_type))
+                fields.append(pa.field(field.name, arrow_type, field.optional))
+            else:
+                raise ResolveException(f"Field is required, and could not be found in the file: {field}")
+
+        return pa.StructArray.from_arrays(arrays=field_arrays, fields=pa.struct(fields))
+
+    def field(self, field: NestedField, _: Optional[pa.Array], field_array: Optional[pa.Array]) -> Optional[pa.Array]:
+        return field_array
+
+    def list(self, list_type: ListType, list_array: Optional[pa.Array], value_array: Optional[pa.Array]) -> Optional[pa.Array]:
+        if list_array is not None and isinstance(list_array, pa.ListArray):
+            return pa.ListArray.from_arrays(list_array.offsets, self.cast_if_needed(list_type.element_field, value_array))
```

Comment:
For maps and lists, the elements/keys/values are visited to project fields in nested structs. Then the same offsets array is used from the partner to reconstruct the final list or map.